src/main/java/org/spongepowered/api/util/blockray/BlockRayFilter.java

```diff
@@ -0,0 +1,260 @@
+
+/**
+ * Represents a filter that determines at what location a {@link BlockRay} should stop. This filter
+ * is called at the boundary of each new location that a {@link BlockRay} passes through in order
+ * to determine whether the ray cast should continue or terminate at that location.
+ *
+ * <p>Any one instance of a {@link BlockRayFilter} should only be run on one path, meaning that the
+ * {@link #start(Location, Vector3d)} method should only be called once. It is not specified that
+ * {@link BlockRayFilter}s have to be stateless, pure functions. They are allowed to keep state
+ * along an individual path, based on the assertion that a single instance is only called on one
+ * path.</p>
+ *
+ * <p>{@link BlockRayFilter}s are most useful for limiting the target block a player is looking
+ * at based on some metric, like transparency, block model, or even distance. The standard
+ * Bukkit-like behavior for finding the target block can be achieved with using
+ * {@link BlockRayFilter#ONLY_AIR}, optionally combined with
+ * {@link BlockRayFilter#maxDistance(double)} to limit the target block to be within some
+ * distance.</p>
+ *
+ * <p>A {@link BlockRayFilter} is modeled as a predicate taking three doubles and a block face,
+ * (in the {@link #shouldContinue(BlockRayHit)} method) instead of using a
+ * regular {@link com.google.common.base.Predicate} taking a tuple. This is to save object
+ * creation, which can become very expensive if an individual ray cast goes through hundreds or
+ * thousands of locations.</p>
+ */
+public abstract class BlockRayFilter {
+
+    /**
+     * Called at the beginning of a ray cast. An instance should perform any setup required
+     * in this method. The default implementation does nothing
+     *
+     * <p>This method can only be called once on any one {@link BlockRayFilter} instance, and is
+     * called at the beginning of a ray cast.</p>
+     *
+     * @param location The starting location of the ray cast
+     * @param direction The direction of the ray cast
+     */
+    public void start(Location location, Vector3d direction) {
+    }
+
+    /**
+     * Called along each step of a ray cast to determine whether the ray cast should continue. A
+     * result value of {@code true} indicates that the ray cast should keep going, while a result
+     * value of {@code false} indicates that the ray cast should stop at the current block
+     * location.
+     *
+     * <p>This method provides the instance of the last block hit, which contains information
+     * about the exact intersection coordinates, the block face hit and the affected block.
+     * If no blocks have been hit yet, this will be the starting position with face of {@link Direction#NONE}.</p>
+     *
+     * @param lastHit The last block hit
+     * @return True to continue ray casting, false to stop
+     * @see BlockRayHit
+     */
+    public abstract boolean shouldContinue(BlockRayHit lastHit);
```

Comment:
But it make for a nicer interface since I can use iterators. Also only one object is created per iteration, as long as you use the getters for the coordinates instead of the vectors.



> On Jun 16, 2015, at 5:53 PM, gratimax <notifications@github.com> wrote:
> 
> In src/main/java/org/spongepowered/api/util/blockray/BlockRayFilter.java:
> 
> > +
> > +    /**
> > +     * Called along each step of a ray cast to determine whether the ray cast should continue. A
> > +     * result value of {@code true} indicates that the ray cast should keep going, while a result
> > +     * value of {@code false} indicates that the ray cast should stop at the current block
> > +     * location.
> > +     *
> > +     * <p>This method provides the instance of the last block hit, which contains information
> > +     * about the exact intersection coordinates, the block face hit and the affected block.
> > +     * If no blocks have been hit yet, this will be the starting position with face of {@link Direction#NONE}.</p>
> > +     *
> > +     * @param lastHit The last block hit
> > +     * @return True to continue ray casting, false to stop
> > +     * @see BlockRayHit
> > +     */
> > +    public abstract boolean shouldContinue(BlockRayHit lastHit);
> The biggest problem I had with doing something like this was that this creates a lot of objects for block rays that span large distances. Even though several objects within block ray hits are lazily-loaded, they still create multiple objects instead, instead of plain xyz and direction as it was previously.
> 
> —
> Reply to this email directly or view it on GitHub.
>